compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -410,93 +410,109 @@ object TypeOps:
     }
   }
 
-  /** An upper approximation of the given type `tp` that does not refer to any symbol in `symsToAvoid`.
+  /** An approximating map that drops NamedTypes matching `toAvoid` and wildcard types. */
+  abstract class AvoidMap(using Context) extends AvoidWildcardsMap:
+    @threadUnsafe lazy val localParamRefs = util.HashSet[Type]()
+
+    def toAvoid(tp: NamedType): Boolean
+
+    /** True iff all NamedTypes on this prefix are static */
+    override def isStaticPrefix(pre: Type)(using Context): Boolean = pre match
+      case pre: NamedType =>
+        val sym = pre.currentSymbol
+        sym.is(Package) || sym.isStatic && isStaticPrefix(pre.prefix)
+      case _ => true
+
+    override def apply(tp: Type): Type = tp match
+      case tp: TermRef
+      if toAvoid(tp) =>
+        tp.info.widenExpr.dealias match {
```

Comment:
Any particular reason to dealias here?

Reply:
This logic was already present before this PR, I assume this is to be able to match a singleton type in the case below when we have a type alias to a singleton type.

Reply:
So we can keep the original in the other case? This is one of the `dealias` calls that causes #14171 and the other one is in `OrderingConstraint.replace`

Reply:
That would be fine yeah, if further dealiasing is needed it will be handled in the recursive call to apply.

Reply:
Although that wouldn't be enough because we could just as well define `type MyType[T] = Nil.type` and get the same issue. This code is a bit confusing to me because `TermRef` is itself a `SingletonType` 🤔

Reply:
Thanks for the explanation.Why does this even trigger in #14171? - the type alias is not local to a block. But regardless we should be careful with dealiasing - if the underlying type wouldn't change it's not desirable to replace an alias because it affects unification. However I think it's happening in so many places that I ran out of steam. I couldn't find any rule as to when `dealias` is applied in the compiler in general.

Reply:
> Why does this even trigger in #14171?

Can't tell from just looking at it.

>  it's not desirable to replace an alias because it affects unification

Yeah there is an attempt at preventing too eager dealiasing in https://github.com/lampepfl/dotty/blob/355d2f6289bba7ee38ab0ac3c389bc7e2a1e39ac/compiler/src/dotty/tools/dotc/core/TypeApplications.scala, but in general it's hard to keep track of them.

Reply:
That link should have been https://github.com/lampepfl/dotty/blob/355d2f6289bba7ee38ab0ac3c389bc7e2a1e39ac/compiler/src/dotty/tools/dotc/core/TypeApplications.scala#L334-L338